### ISSUE TITLE
flux reconcile on helmrelease leaves erroneously modified object in bad state

### ISSUE THREAD

**Issue by scott-grimes** [fluxcd/helm-controller#267]
### Describe the bug

A HelmRelease contains a cronjob as part of a release. If we edit that cronjob in k8s and then run `flux reconcile -n <namespace> helmrelease <hr name> --with-source`, the edited cronjob is **not** replaced with the correct cronjob.

### To Reproduce

1) HelmRelease is READY, and helm shows STATUS as deployed
```
$ kubectl get hr
NAME             READY   STATUS                             AGE
blerg           True    Release reconciliation succeeded   9d
$ helm ls
NAME          	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART               	APP VERSION
blerg        	prod     	224     	2021-05-14 19:30:24.051846927 +0000 UTC	deployed	blerg-0.2.0        	1.0
```

2) I can see my cronjob spec in the cluster, it is exactly what helm says it should be
```
$ kubectl get cronjob myjob -o yaml
...
spec:
  containers:
  - args:
    - /bin/bash
    - /app/scripts/run
    ...
###
$ helm get all blerg > manifests.yaml
# grep for CronJob
...
spec:
  containers:
  - args:
    - /bin/bash
    - /app/scripts/run
    ...
```

3) Manually edit the cronjob, adding a "command" section to the container spec 
```
$ kubectl edit cronjob myjob
...
spec:
  containers:
  - args:
    - /bin/bash
    - /app/scripts/run
    command:
    - /bin/bash
    - -c
    ...
```

4) Reconcile, which should replace the edited job with the correct job
```
$ flux reconcile -n prod helmrelease blerg --with-source
► annotating GitRepository myrepo in flux-system namespace
✔ GitRepository annotated
◎ waiting for GitRepository reconciliation
✔ GitRepository reconciliation completed
✔ fetched revision master/<sha>
► annotating HelmRelease blerg in prod namespace
✔ HelmRelease annotated
◎ waiting for HelmRelease reconciliation
✔ HelmRelease reconciliation completed
✔ reconciled revision 0.2.0
```

5) I can see reconciliation working, the update hooks are running, etc 
```
$ kubectl get hr 
NAME             READY     STATUS                             AGE
blerg           Unknown   Reconciliation in progress   
...
$ helm ls -a
NAME          	NAMESPACE	REVISION	UPDATED                                	STATUS         	CHART               	APP VERSION
blerg        	prod     	225     	2021-05-14 19:43:50.894886714 +0000 UTC	pending-upgrade	blerg-0.2.0        	1.0 
```

6) I wait for helm to complete. HelmRelease is READY, and helm shows STATUS as deployed
```
$ kubectl get hr
NAME             READY   STATUS                             AGE
blerg           True    Release reconciliation succeeded   9d
$ helm ls
NAME          	NAMESPACE	REVISION	UPDATED                                	STATUS  	CHART               	APP VERSION
blerg        	prod     	225     	2021-05-14 19:43:50.894886714 +0000 UTC	deployed	blerg-0.2.0        	1.0        
```

7) I can now look at the cronjob and see that my modification has remained in place, even though it is not supposed to be there. 
```
$ kubectl get cronjob myjob -o yaml
...
spec:
  containers:
  - args:
    - /bin/bash
    - /app/scripts/run
    command:
    - /bin/bash
    - -c
    ...
###
$ helm get all blerg > manifests.yaml
# grep for CronJob
...
spec:
  containers:
  - args:
    - /bin/bash
    - /app/scripts/run
    # no "command" section
    ...
```

### Expected behavior

The invalid cronjob should have been replaced with helms manifest

### Additional context

- Kubernetes version:
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-14T05:14:17Z", GoVersion:"go1.15.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.12-eks-7684af", GitCommit:"7684af4ac41370dd109ac13817023cb8063e3d45", GitTreeState:"clean", BuildDate:"2020-10-20T22:57:40Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
- Git provider:
Github

Below please provide the output of the following commands:

```cli
flux --version
flux version 0.11.0
flux check
► checking prerequisites
✗ flux 0.11.0 <0.13.4 (new version is available, please upgrade)
✔ kubectl 1.20.2 >=1.18.0-0
✔ Kubernetes 1.17.12-eks-7684af >=1.16.0-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.9.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.10.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.11.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.10.0
✔ all checks passed
kubectl -n <namespace> logs deploy/source-controller
{"level":"info","ts":"2021-05-14T19:53:48.826Z","logger":"controller.gitrepository","msg":"Reconciliation finished in 350.449894ms, next run in 1m0s","reconciler group":"source.toolkit.fluxcd.io","reconciler kind":"GitRepository","name":"...","namespace":"..."}
kubectl -n <namespace> logs deploy/kustomize-controller
{"level":"info","ts":"2021-05-14T19:54:37.116Z","logger":"controller.kustomization","msg":"Kustomization applied in 456.616953ms","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"...","namespace":"...","output":{"<all unchanged>}}
{"level":"info","ts":"2021-05-14T19:54:37.141Z","logger":"controller.kustomization","msg":"Reconciliation finished in 640.896612ms, next run in 3m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"...","namespace":"...","revision":"master/<sha>"}
```

**Comment by mmckane** [comment]
I am seeing this as well on flux 0.20 it seems like diffing with helmreleases is broken. It also seems that if I delete a resource manually flux doesn't seem to be picking up that it should put it back and won't until the chart version is modified or the git repo is updated when using reconcileStrategy: Revision.

**Comment by zaggash** [comment]
> I am seeing this as well on flux 0.20 it seems like diffing with helmreleases is broken. It also seems that if I delete a resource manually flux doesn't seem to be picking up that it should put it back and won't until the chart version is modified or the git repo is updated when using reconcileStrategy: Revision.

I experience the same with deleted objects from a HelmRelease, lets say an ingress.
If I delete it, it is not recreated on a reconcile.

**Comment by monotek** [comment]
Same here.
Imho helm-controller should act as "helm upgrade" command as this updates the resources to it's desired state.

**Comment by kingdonb** [comment]
> If I delete it, it is not recreated on a reconcile.

@zaggash @monotek This is expected behavior. To understand why, you must understand how Helm resolves drifts, (and what obstacles it has placed in its own way with respect to that story)

* https://github.com/fluxcd/helm-controller/issues/186

**Tl;dr**: helm controller doesn't reconcile drift in the cluster, except when it is performing an actual upgrade (and that drift in the cluster wouldn't currently be able to trigger any upgrade, because for Helm Controller it's perfectly indistinguishable from any side-effects that might have been the intended result of any post-render or lifecycle hook behavior in the chart.)

I am not certain if these reports are the same as the original report, but the behavior that both @zaggash and @monotek are describing are already explained by #186 as I'll elaborate further on below. (This is a complicated story, so save yourself and look the other way, unless the tl;dr above has made you hungry to understand what the problem is and why this issue can't easily be resolved.

---

Helm uses a 3-way merge to incorporate: (1) the old manifest, (2) its live state on the cluster, and (3) the new manifest.

Helm also provides various opportunities for the installing agent to manipulate the state before, around and after install, both within the chart ([lifecycle hooks](https://helm.sh/docs/topics/charts_hooks/#the-available-hooks)) that are permitted to be non-deterministic (and so may have outputs that are not declaratively formed), and also outside of the chart, with [Post Render](https://fluxcd.io/docs/components/helm/helmreleases/#post-renderers) behavior, that outside of Flux, would also be capable of non-deterministic behavior.

Simplified: The installer can create secrets on the first run (`helm install`), that won't be generated again during a second run (`helm upgrade`).

Though Flux makes the postrender behavior declarative, it's still running the Helm code under the hood, which does not assume declarativeness, and that means we can't assume anything that Helm itself would not have assumed or guaranteed.

Every upgrade produces a new `release` artifact, in the form of a secret of the type: `helm.sh/release.v1`

So when you have just installed for the first time and you have a handful of Helm charts, your cluster-wide secrets may look like this:

```
chartmuseum           sh.helm.release.v1.chartmuseum-chartmuseum.v1                  helm.sh/release.v1                    1      4d13h
deis                  sh.helm.release.v1.deis-hephy.v1                               helm.sh/release.v1                    1      4d13h
harbor                sh.helm.release.v1.harbor.v1                                   helm.sh/release.v1                    1      4d13h
ingress-nginx         sh.helm.release.v1.ingress-internal.v1                         helm.sh/release.v1                    1      4d13h
ingress-nginx         sh.helm.release.v1.ingress-public.v1                           helm.sh/release.v1                    1      4d13h
kube-oidc-proxy       sh.helm.release.v1.kube-oidc-proxy.v1                          helm.sh/release.v1                    1      4d11h
minio-lusers          sh.helm.release.v1.minio-lusers-minio.v1                       helm.sh/release.v1                    1      4d13h
minio-stage           sh.helm.release.v1.minio-stage-minio.v1                        helm.sh/release.v1                    1      4d13h
monitoring            sh.helm.release.v1.kube-prometheus-stack.v1                    helm.sh/release.v1                    1      2d22h
monitoring            sh.helm.release.v1.kube-prometheus-stack.v2                    helm.sh/release.v1                    1      2d22h
openvpn               sh.helm.release.v1.openvpn.v1                                  helm.sh/release.v1                    1      4d13h
```

Three-way diffs are complicated, even more complicated than three-way merges. Now imagine instead of comparing the inputs to the outputs, Helm controller just runs an upgrade each time, producing a new revision and a new versioned secret.

The two secrets shown here for `v1` and `v2` revisions in the `monitoring` namespace will quickly fall off the edge, as new release secrets are generated upon each upgrading reconcile, and we've effectively made Helm's `rollback` feature useless. Helm controller can only infer from the changing state in any inputs, that it should perform an upgrade.

---

I am virtually certain this does not have anything to do with the original issue, sorry for contributing to the noise, but this is a hard issue to explain, and I'm sure it comes up all the time. Though I am afraid I have not quite understood the cronjob issue that started this thread, doing my best to steer the discussion back on topic and find out if any fix is possible, (or if we might have already fixed it somehow without noticing)
